File: src/main/java/kitchenpos/menu/tobe/application/MenuService.java
package kitchenpos.menu.tobe.application;

import kitchenpos.common.exception.MenuException;
import kitchenpos.common.exception.MenuNotFoundException;
import kitchenpos.common.tobe.Profanities;
import kitchenpos.menu.tobe.application.dto.request.MenuCreateRequest;
import kitchenpos.menu.tobe.application.dto.request.MenuPriceChangeRequest;
import kitchenpos.menu.tobe.domain.menu.*;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;

import static kitchenpos.common.exception.ErrorCode.*;

@Service
public class MenuService {
    private final MenuRepository menuRepository;
    private final Profanities profanities;
    private final MenuValidator menuValidator;

    public MenuService(
            final MenuRepository menuRepository,
            final Profanities profanities,
            final MenuValidator menuValidator
    ) {
        this.menuRepository = menuRepository;
        this.profanities = profanities;
        this.menuValidator = menuValidator;
    }

    @Transactional
    public Menu create(final MenuCreateRequest request) throws MenuException {
        return menuRepository.save(Menu.of(
                MenuName.from(request.name(), profanities),
                MenuPrice.from(request.price()),
                request.menuGroupId(),
                MenuDisplayStatus.from(request.displayed()),
                MenuProducts.from(request.menuProducts()),
                menuValidator
        ));
    }

    @Transactional
    public Menu changePrice(final UUID menuId, final MenuPriceChangeRequest request) {
        final Menu menu = menuRepository.findById(menuId)
                .orElseThrow(() -> new MenuNotFoundException(MENU_NOT_FOUND));
        menu.changeMenuPrice(request.price(), menuValidator);
        return menu;
    }

    @Transactional
    public Menu display(final UUID menuId) {
        final Menu menu = menuRepository.findById(menuId)
                .orElseThrow(() ->
                        new MenuNotFoundException(MENU_NOT_FOUND)
                );
        menu.show(menuValidator);
        return menu;
    }

    @Transactional
    public Menu hide(final UUID menuId) {
        final Menu menu = menuRepository.findById(menuId)
                .orElseThrow(() -> new MenuNotFoundException(MENU_NOT_FOUND));
        menu.hide();
        return menu;
    }

    @Transactional(readOnly = true)
    public List<Menu> findAll() {
        return menuRepository.findAll();
    }
}

File: src/main/kotlin/kitchenpos/menu/tobe/application/MenuService.kt
package kitchenpos.menu.tobe.application

import java.util.*
import kitchenpos.menu.tobe.application.dto.CreateMenuReq
import kitchenpos.menu.tobe.application.dto.MenuResp
import kitchenpos.menu.tobe.domain.Menu
import kitchenpos.menu.tobe.domain.MenuGroupRepository
import kitchenpos.menu.tobe.domain.MenuName
import kitchenpos.menu.tobe.domain.MenuNamePolicy
import kitchenpos.menu.tobe.domain.MenuPrice
import kitchenpos.menu.tobe.domain.MenuProduct
import kitchenpos.menu.tobe.domain.MenuProducts
import kitchenpos.menu.tobe.domain.MenuRepository
import kitchenpos.menu.tobe.domain.ProductClient
import kitchenpos.product.tobe.application.dto.ChangeProductPriceReq
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service(value = "tobeMenuService")
class MenuService(
    private val menuRepository: MenuRepository,
    private val menuGroupRepository: MenuGroupRepository,
    private val menuNamePolicy: MenuNamePolicy,
    private val productClient: ProductClient,
) {
    @Transactional
    fun create(request: CreateMenuReq): MenuResp {
        val menuGroup = menuGroupRepository.findById(request.menuGroupId)
            .orElseThrow { NoSuchElementException("존재하지 않는 메뉴 그룹입니다.") }
        val productInfos = productClient.getProducts(request.menuProducts.map { it.productId });

        val menu = Menu(
            productInfos = productInfos,
            menuName = MenuName(menuNamePolicy = menuNamePolicy, name = request.name),
            menuPrice = MenuPrice(request.price.toBigDecimal()),
            menuGroup = menuGroup,
            menuDisplay = request.display,
            menuProducts = MenuProducts(request.menuProducts.map { menuProductReq ->
                productInfos[menuProductReq.productId]
                    ?: throw NoSuchElementException("존재하지 않는 상품입니다.")
                MenuProduct(productId = menuProductReq.productId, quantity = menuProductReq.quantity)
            })
        )
        menuRepository.save(menu)
        return MenuResp.of(menu)
    }

    @Transactional
    fun changePrice(menuId: UUID, request: ChangeProductPriceReq) {
        val menu = menuRepository.findById(menuId)
            .orElseThrow { NoSuchElementException("존재하지 않는 메뉴입니다.") }
        val productInfos = productClient.getProducts(menu.productIds())
        menu.changePrice(productInfos, MenuPrice(request.price))
        menuRepository.save(menu)
    }

    @Transactional
    fun display(menuId: UUID) {
        val menu = menuRepository.findById(menuId)
            .orElseThrow { NoSuchElementException("존재하지 않는 메뉴입니다.") }
        val productInfos = productClient.getProducts(menu.productIds())
        menu.display(productInfos)
    }

    @Transactional
    fun notDisplay(menuId: UUID) {
        val menu = menuRepository.findById(menuId)
            .orElseThrow { NoSuchElementException("존재하지 않는 메뉴입니다.") }
        menu.notDisplay()
    }
}
